tools/check_python_h_first.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python
+"""Check that Python.h is included before any stdlib headers.
+
+May be a bit overzealous, but it should get the job done.
+"""
+import argparse
+import os.path
+import re
+import sys
+
+HEADER_PATTERN = re.compile(
+    r'^\s*#\s*include\s*[<"]((?:\w+/)*\w+(?:\.h[hp+]{0,2})?)[>"]\s*$'
+)
+
+PYTHON_INCLUDING_HEADERS = [
+    "Python.h",
+    "numpy/arrayobject.h",
+    "numpy/ndarrayobject.h",
+    "numpy/npy_common.h",
+    "numpy/npy_math.h",
+    "numpy/random/distributions.h",
+    "pybind11/pybind11.h",
+    # Python-including headers the sort doesn't pick up
+    "ni_support.h",
+]
+
+PARSER = argparse.ArgumentParser(description=__doc__)
+PARSER.add_argument("file_list", nargs="+", type=str)
+
+
+def check_python_h_included_first(name_to_check: str) -> int:
+    """Check that the passed file includes Python.h first if it does at all.
+
+    Perhaps overzealous, but that should work around concerns with
+    recursion.
+
+    Parameters
+    ----------
+    name_to_check : str
+        The name of the file to check.
+
+    Returns
+    -------
+    int
+        The number of headers before Python.h
+    """
+    included_python = False
+    included_non_python_header = []
+    warned_python_construct = False
+    basename_to_check = os.path.basename(name_to_check)
+    in_comment = False
+    with open(name_to_check) as in_file:
+        for i, line in enumerate(in_file, 1):
+            # Very basic comment parsing
+            # Assumes /*...*/ comments are on their own lines
+            if "/*" in line:
+                if "*/" not in line:
+                    in_comment = True
+                # else-branch could use regex to remove comment and continue
+                continue
+            if in_comment:
+                if "*/" in line:
+                    in_comment = False
+                continue
+            match = re.match(HEADER_PATTERN, line)
```

Comment:
Just a drive by comment, but the code runs more than twice as fast for me locally on our codebase if I use the compiled regex directly:

```diff
--- a/tools/check_python_h_first.py
+++ b/tools/check_python_h_first.py
@@ -62,7 +62,7 @@ def check_python_h_included_first(name_to_check: str) -> int:
                 if "*/" in line:
                     in_comment = False
                 continue
-            match = re.match(HEADER_PATTERN, line)
+            match = HEADER_PATTERN.match(line)
             if match:
                 this_header = match.group(1)
                 if this_header in PYTHON_INCLUDING_HEADERS:
```

Apart from that, I didn't see any issues from playing with it for a few minutes and reading the diff.

Reply:
```suggestion
            match =HEADER_PATTERN.match(line)
```
Was not aware I could do that.  Done.